implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java

```diff
@@ -158,57 +310,65 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
-            }
-        }
-
-        return status;
-    }
-
-    private HealthCheckResponse.State fillCheck(HealthCheck check, JsonArrayBuilder results,
+    private HealthCheckResponse.State handleResponse(HealthCheckResponse response, JsonArrayBuilder results,
             HealthCheckResponse.State globalOutcome) {
-        if (check == null) {
-            return globalOutcome;
-        }
-        JsonObject each = jsonObject(check);
-        results.add(each);
+        JsonObject responseJson = jsonObject(response);
+        results.add(responseJson);
+
         if (globalOutcome == HealthCheckResponse.State.UP) {
-            String status = each.getString("status");
+            String status = responseJson.getString("status");
             if (status.equals("DOWN")) {
                 return HealthCheckResponse.State.DOWN;
             }
         }
+
         return globalOutcome;
     }
 
-    private JsonObject jsonObject(HealthCheck check) {
-        try {
-            return jsonObject(check.call());
-        } catch (RuntimeException e) {
-            // Log Stacktrace to server log so an error is not just in Health Check response
-            HealthLogging.log.healthCheckError(e);
-
-            HealthCheckResponseBuilder response = HealthCheckResponse.named(check.getClass().getName()).down();
-
-            if (null != uncheckedExceptionDataStyle) {
-                switch (uncheckedExceptionDataStyle) {
-                    case ROOT_CAUSE:
-                        response.withData(ROOT_CAUSE, getRootCause(e).getMessage());
-                        break;
-                    case STACK_TRACE:
-                        response.withData(STACK_TRACE, getStackTrace(e));
-                        break;
-                    default:
-                        // don't add anything
-                }
-            }
+    private Uni<HealthCheckResponse> callAsync(AsyncHealthCheck asyncHealthCheck) {
+        // must be a supplier to be reinvoked on every subcribtion, don't change to method reference
+        return withRecovery(asyncHealthCheck.getClass().getName(),
+                Uni.createFrom().deferred(() -> {
+                    try {
+                        Uni<HealthCheckResponse> call = asyncHealthCheck.call();
+                        if (call == null) {
+                            throw HealthMessages.msg.healthCheckNull();
+                        }
+                        return call;
+                    } catch (IllegalArgumentException ex) {
+                        if (ex.getMessage().matches(".*supplier.*null.*")) {
+                            throw HealthMessages.msg.healthCheckNull();
+                        }
+                        throw ex;
+                    }
+                }));
+    }
+
+    private Uni<HealthCheckResponse> callSync(HealthCheck healthCheck) {
+        // must be a supplier to be reinvoked on every subcribtion, don't change to method reference
```

Comment:
it will work the same way with a method reference. What should be avoided is: `Uni.createFrom().item(healthCheck.call());`

Reply:
yes, this is exactly what my IDE was trying to change :) thus the comment.

Reply:
Uni.createFrom().item(healthCheck.call()); != Uni.createFrom().item(HealthCheck::call);

The second one would use the `item` method using the supplier.

Reply:
I'll be dammed. Surely something I've overlooked coding before sleep :)